### PR TITLE
Integrate NeoFS into oracle module

### LINKED ISSUE #1473: NeoFS integration into oracle module
Let's experiment a little and after oracle/module settles make a separate branch for NeoFS based on top of it (not merging it into master). We'll follow neo-project/neo#1270 and neo-project/neo#1256 URI schemes.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1473 .

### CODE REVIEW

File: pkg/oracle/neofs/neofs.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+// 	Addresses are queried sequentially using the same context.
+func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string) ([]byte, error) {
+	if u.Scheme != URIScheme {
+		return nil, ErrInvalidScheme
+	}
+	ps := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
+	if len(ps) == 0 {
+		return nil, ErrMissingObject
+	}
+	spew.Dump(ps)
```
Comment (roman-khimov) [inline pkg/oracle/neofs/neofs.go:58]: No spewing here, please.

File: pkg/oracle/neofs/neofs.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+	rawCID, err := base58.Decode(u.Hostname())
+	if err != nil || len(rawCID) != containerIDSize {
+		return nil, ErrInvalidContainer
+	}
+	rawOID, err := base58.Decode(ps[0])
+	if err != nil || len(rawCID) != objectIDSize {
```
Comment (roman-khimov) [inline pkg/oracle/neofs/neofs.go:65]: s/rawCID/rawOID/

File: pkg/oracle/neofs/neofs.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	r.SetLength(length)
+	return c.ObjectPayloadRangeData(ctx, new(client.RangeDataParams).WithAddress(addr).WithRange(r))
+}
+
+func getHeader(ctx context.Context, c *client.Client, addr *object.Address) ([]byte, error) {
+	obj, err := c.GetObjectHeader(ctx, new(client.ObjectHeaderParams).WithAddress(addr))
+	if err != nil {
+		return nil, err
+	}
+	return obj.GetPayload(), nil // TODO
```
Comment (roman-khimov) [inline pkg/oracle/neofs/neofs.go:135]: So how does it differ from `getPayload`?

File: pkg/oracle/neofs/neofs.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+	// containerIDSize is the size of container id in bytes.
+	containerIDSize = sha256.Size
+	// objectIDSize is the size of container id in bytes.
+	objectIDSize = sha256.Size
+	// rangeSep is a separator between offset and length.
+	rangeSep = '|'
+
+	rangeCmd  = "range"
+	headerCmd = "header"
```
Comment (roman-khimov) [inline pkg/oracle/neofs/neofs.go:32]: `hashCmd`?

File: pkg/oracle/neofs/neofs_test.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	u, err := url.ParseRequestURI(addr)
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	priv, err := keys.NewPrivateKey()
+	require.NoError(t, err)
+
+	data, err := Get(ctx, priv, u, "192.168.130.71:8080")
```
Comment (roman-khimov) [inline pkg/oracle/neofs/neofs_test.go:26]: I don't think it's gonna work for me.

File: pkg/oracle/request.go [hunk 6]
```diff
@@ -127,6 +130,18 @@ func (o *Oracle) processRequest(priv *keys.PrivateKey, req request) error {
 		default:
 			resp.Code = transaction.Error
 		}
+	} else if err == nil && u.Scheme == neofs.URIScheme {
+		ctx, cancel := context.WithTimeout(context.Background(), o.MainCfg.RequestTimeout)
+		defer cancel()
+		index := incTx.attempts % len(o.MainCfg.NeoFSNodes)
```
Comment (roman-khimov) [inline pkg/oracle/request.go:136]: This makes `NeoFSNodes[0]` serve all requests most of the time, but we need to spread the load, so you're better using request ID plus attempts here.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/oracle/neofs/neofs.go (hunk 1 above)
Comment (fyrchik) [reply pkg/oracle/neofs/neofs.go:58]: I will remove this after finish testing.

File: pkg/oracle/neofs/neofs.go (hunk 2 above)
Comment (fyrchik) [reply pkg/oracle/neofs/neofs.go:65]: Fixed.

File: pkg/oracle/neofs/neofs.go (hunk 1 above)
Comment (fyrchik) [reply pkg/oracle/neofs/neofs.go:58]: Fixed.

File: pkg/oracle/neofs/neofs.go (hunk 4 above)
Comment (fyrchik) [reply pkg/oracle/neofs/neofs.go:32]: Fixed.

File: pkg/oracle/request.go (hunk 6 above)
Comment (fyrchik) [reply pkg/oracle/request.go:136]: Fixed.

File: pkg/oracle/neofs/neofs.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/oracle/neofs/neofs.go]: It's still there in tests and it adds an additional dependency.

File: pkg/oracle/neofs/neofs.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	return c.ObjectPayloadRangeData(ctx, new(client.RangeDataParams).WithAddress(addr).WithRange(r))
+}
+
+func getHeader(ctx context.Context, c *client.Client, addr *object.Address) ([]byte, error) {
+	obj, err := c.GetObjectHeader(ctx, new(client.ObjectHeaderParams).WithAddress(addr))
+	if err != nil {
+		return nil, err
+	}
+	return obj.GetPayload(), nil // TODO json header
```
Comment (roman-khimov) [inline pkg/oracle/neofs/neofs.go:124]: Waiting for it.

File: pkg/oracle/neofs/neofs.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	containerID := container.NewID()
+	containerID.SetSHA256(cid)
+
+	copy(oid[:], rawOID)
+	objectID := object.NewID()
+	objectID.SetSHA256(oid)
+
+	objectAddr := object.NewAddress()
+	objectAddr.SetContainerID(containerID)
+	objectAddr.SetObjectID(objectID)
```
Comment (alexvanin) [inline pkg/oracle/neofs/neofs.go:82]: With latest neofs-api-go you can simplify parsing a bit:

```go
	containerID := container.NewID()
	err = containerID.Parse(u.Hostname())
 	if err != nil {
		return nil, ErrInvalidContainer
	}

	objectID := object.NewID()
	err = objectID.Parse(ps[0])
 	if err != nil {
		 return nil, ErrInvalidObject
	}

	objectAddr := object.NewAddress()
	objectAddr.SetContainerID(containerID)
	objectAddr.SetObjectID(objectID)
```

Comment (alexvanin) [reply pkg/oracle/neofs/neofs.go:82]: Still relevant: we can replace base58 encoding and copying SHA256 data with `containerID.Parse(u.Hostname())` for container and  `objectID.Parse(ps[0])` for object.

Comment (roman-khimov) [reply pkg/oracle/neofs/neofs.go:82]: @fyrchik?

Comment (fyrchik) [reply pkg/oracle/neofs/neofs.go:82]: Fixed.

**Comment by fyrchik** [comment]
Update:
1. Made config compatible with C# node (we can replace single node with a list after https://github.com/go-yaml/yaml/issues/100 or change C# node directly).
2. Added tests for parsing routines.

File: pkg/config/oracle_config.go [hunk 9]
```diff
@@ -4,13 +4,20 @@ import "time"
...
+	RefreshInterval       time.Duration      `yaml:"RefreshInterval"`
+	MaxConcurrentRequests int                `yaml:"MaxConcurrentRequests"`
+	RequestTimeout        time.Duration      `yaml:"RequestTimeout"`
+	ResponseTimeout       time.Duration      `yaml:"ResponseTimeout"`
+	UnlockWallet          Wallet             `yaml:"UnlockWallet"`
+}
+
+// NeoFSConfiguration is a config for the NeoFS service.
+type NeoFSConfiguration struct {
+	EndPoint string `yaml:"EndPoint"`
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/config/oracle_config.go]: Restored that.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**